subprojects/core/src/main/java/org/gradle/internal/operations/PreserveBuildOperationIdentifierRunner.java

```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.operations;
+
+public class PreserveBuildOperationIdentifierRunner implements Runnable {
+    private final Runnable delegate;
+    private final Object buildOperationId;
+
+    private PreserveBuildOperationIdentifierRunner(Runnable delegate, Object buildOperationId) {
+        this.delegate = delegate;
+        this.buildOperationId = buildOperationId;
+    }
+
+    @Override
+    public void run() {
+        BuildOperationIdentifierRegistry.setCurrentOperationIdentifier(buildOperationId);
+        try {
+            delegate.run();
+        } finally {
+            BuildOperationIdentifierRegistry.clearCurrentOperationIdentifier();
+        }
+    }
+
+    public static Runnable wrap(Runnable delegate) {
```

Comment:
:x: I don't quite see why we need the static method. Couldn't the constructor just look like this?

```
public PreserveBuildOperationIdentifierRunner(Runnable delegate) {
    this.delegate = delegate;
    buildOperationId = BuildOperationIdentifierRegistry.getCurrentOperationIdentifier();
}
```